core/src/main/java/org/apache/calcite/plan/ConventionTraitDef.java

```diff
@@ -234,10 +234,10 @@ private ConversionData getConversionData(RelOptPlanner planner) {
       return pathMap;
     }
 
-    public List<Convention> getShortestPath(
+    public int getShortestDistance(
```

Comment:
Can you also fix the getPaths() to return shortest paths first? This is to make sure we choose the shortest path during convert.

Reply:
IMO, it would be an overkill to call getPaths() only to get the shortest path. 
To support this scenario, I have restored the getShortestPath API, and implement it with BFS. It should be more efficient than the Dijkstra and Floyd algorihtms.

Reply:
IMHO this is not the ideal approach, now we would be re-computing everything time something (shortest path) that was previously computed only once and returned in O(1).

If we take a few steps back, I think it is clear that we require both pieces of information (shortesPath & distance), to be provided ASAP from `FrozenGraph`. Then why not just pre-computing both in `Graphs#makeImmutable` and storing both of them in `FrozenGraph`, so that we guarantee that `getShortestPath` & `getShortestDistance` are executed in O(1)?
I think we could keep track of shortestPath and distance in `Graphs#makeImmutable`, somehow combining the old approach with the newly proposed approach, either keeping two maps:
```
Map<Pair<V, V>, List<V>> shortestPaths
Map<Pair<V, V>, int[]> shortestDistances
```
Or a single map with the combination of both as value:
```
Map<Pair<V, V>, Pair<List<V>, Integer>> shortestPathsAndDistances
```
Then we would pass this information as a parameter for FrozeGraph constructor, and we would have shortesPath & distance pre-computed from the beginning.
I'm not sure if what I say makes sense or it is an overkill. What do you guys think?

Reply:
@liyafan82 I am not saying getPaths() to return shortest paths. What I mean is you should put shortest path at the front of list, so during convert() we always choose the shortest converted path if possible. This only requires a sort after generating all possible path.

@rubenada I don't think we need to cache shortest path. This is only one time used and there's no benefit to cache it.

Reply:
@xndai I have revised the getPaths() method accordingly. Thank you.

@rubenada Thanks a lot for your suggestion. I tend to agree with @xndai. If getting shortest paths is not a frequently used operation, there is no need to store all pairs of shortest paths. Here, we preserve the getShortestPath API, mainly for the sake of backward compatibility.